tests/performance/valgrind_instrumentation.h

```diff
@@ -0,0 +1,143 @@
+ *            --callgrind-out-file=callgrind.out ./executable
+ * @endcode
+ */
+namespace callgrind_wrapper
+{
+  /**
+   * Reset the current instruction counter to zero and start callgrind
+   * instrumentation.
+   */
+  DEAL_II_ALWAYS_INLINE
+  inline void
+  start_instrumentation()
+  {
+    /*
+     * Verify that we are running under valgrind and that valgrind is
+     * called with the right parameters.
+     */
+    static bool is_initialized = []() {
+      // Is this executable running under valgrind supervision?
+      if (RUNNING_ON_VALGRIND != 1)
+        return false;
+
+      // Start and stop valgrind instrumentation
+      CALLGRIND_STOP_INSTRUMENTATION;
+      std::remove("callgrind.out");
+      CALLGRIND_ZERO_STATS;
+      CALLGRIND_START_INSTRUMENTATION;
+      const unsigned my_pid = getpid(); // do something useful
+      CALLGRIND_DUMP_STATS_AT("callgrind-wrapper-token");
+      CALLGRIND_STOP_INSTRUMENTATION;
+
+      std::ifstream      callgrind_output("callgrind.out");
+      std::string        token;
+      unsigned long long cycles      = 0ull;
+      bool               found_token = false;
+      unsigned long      found_pid   = 0;
+
+      while (callgrind_output)
+        {
+          callgrind_output >> token;
+          if (token == "pid:")
+            callgrind_output >> found_pid;
+          if (token == "desc:")
+            {
+              std::string line;
+              std::getline(callgrind_output, line);
+              if (line == " Trigger: Client Request: callgrind-wrapper-token")
+                found_token = true;
+            }
+          if (token == "summary:")
+            callgrind_output >> cycles;
+        }
+
+      if (!found_token)
+        return false;
+
+      if (found_pid != my_pid)
+        return false;
+
+      if (cycles == 0)
+        return false;
+
+      return true;
+    }();
+
+    AssertThrow(
+      is_initialized,
+      dealii::ExcMessage(
+        "callgrind_wrapper::start_instrumentation() can only be called when "
+        "the executable is run via \"valgrind --tool=callgrind -q "
+        "--combine-dumps=yes --instr-atstart=no "
+        "--callgrind-out-file=callgrind.out ./executable\""));
+
+    CALLGRIND_ZERO_STATS;
+    CALLGRIND_START_INSTRUMENTATION;
+  }
+
+
+  /**
+   * Stop callgrind instrumentation and return the number of instructions
+   * executed since the last start_instrumentation() call.
+   */
+  DEAL_II_ALWAYS_INLINE
+  inline unsigned long long
+  stop_instrumentation()
+  {
+    CALLGRIND_DUMP_STATS;
+    CALLGRIND_STOP_INSTRUMENTATION;
+
+    std::ifstream      callgrind_output("callgrind.out");
+    std::string        token;
+    unsigned long long cycles = 0ull;
+
+    while (callgrind_output)
+      {
+        callgrind_output >> token;
+        if (token == "summary:")
+          callgrind_output >> cycles;
+      }
+
+    std::remove("callgrind.out");
+    return cycles;
+  }
+
```

Comment:
What you really want in this file is a function of the form
```
template <typename Func>
inline std::uint64_t
count_cycles (Func &&f)
{
  start_instrumentation();
  f();
  return stop_instrumentation();
}
```
so that in the performance tests, you can replace
```
  callgrind_wrapper::start_instrumentation();
  setup_system();
  cycle_count["setup_system"] = callgrind_wrapper::stop_instrumentation();
```
by
```
  cycle_count["setup_system"] = callgrind_wrapper::count_cycles([this]() { setup_system(); });
```

Reply:
Let's make such a function call wrapper and/or possible changes to the API part of a future pull request :smiley: